File: ruby/grid.rb
def set_piece(grid, i, j, piece, char)
  piece[:coordinates].each do |coordinate|
    grid[i + coordinate[0]][j + coordinate[1]] = char
  end
  return grid
end

def print_grid(grid)
  puts "Vous avez placé #{grid.flatten.join.chars.uniq.size - 1} pièces."
  puts ""
  grid.each { |line| puts line.join }
  puts ""
end

def minimal_grid_size(n)
  grid_size = 2
  while grid_size ** 2 < n * 4
    grid_size += 1
  end
  return grid_size
end

def place_piece(grid, piece, i, j)
  piece[:coordinates].each do |coordinate|
    if grid[i + coordinate[0]][j + coordinate[1]] != "."
      return false
    end
  end
  set_piece(grid, i, j, piece, piece[:letter])
  return true
end

File: SRC/grid.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   grid.c                                             :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: fwuensch <fwuensch@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/01/19 18:35:03 by fwuensch          #+#    #+#             */
/*   Updated: 2019/01/24 22:22:54 by fwuensch         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "fillit.h"

void		print_grid(char **grid, int grid_size)
{
	int		i;
	int		j;

	i = 0;
	while (i < grid_size)
	{
		j = 0;
		while (j < grid_size)
		{
			ft_putchar(grid[i][j]);
			j++;
		}
		ft_putchar('\n');
		i++;
	}
}

char		**create_grid(void)
{
	char	**grid;
	int		i;

	grid = (char**)malloc(sizeof(char*) * GRID_MAX);
	if (grid == NULL)
		return (NULL);
	i = 0;
	while (i < GRID_MAX)
	{
		grid[i] = (char*)malloc(sizeof(char) * GRID_MAX);
		if (grid[i] == NULL)
			return (NULL);
		i++;
	}
	return (grid);
}

void		set_piece(char **grid, int i, int j, t_piece piece)
{
	int		x;

	x = 0;
	while (x < 4)
	{
		grid[i + piece.coord[x][0]][j + piece.coord[x][1]] = piece.letter;
		x++;
	}
}

void		remove_piece(char **grid, int i, int j, t_piece piece)
{
	int		x;

	x = 0;
	while (x < 4)
	{
		grid[i + piece.coord[x][0]][j + piece.coord[x][1]] = '.';
		x++;
	}
}

int			place_piece(char **grid, t_piece piece, int i, int j)
{
	int		x;

	x = 0;
	while (x < 4)
	{
		if (grid[i + piece.coord[x][0]][j + piece.coord[x][1]] != '.')
			return (0);
		x++;
	}
	set_piece(grid, i, j, piece);
	return (1);
}
